sdks/java/core/src/main/java/org/apache/beam/sdk/transforms/RegexTransform.java

```diff
@@ -0,0 +1,318 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.transforms;
+
+import org.apache.beam.sdk.values.KV;
+import org.apache.beam.sdk.values.PCollection;
+
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+/**
+ * {@code PTransorm}s to use Regular Expressions to process elements in a
+ * {@link PCollection}.
+ *
+ * <p>
+ * {@link RegexTransform#matches(String, int)} can be used to see if an entire line matches
+ * a Regex. {@link RegexTransform#matchesKV(String, int, int)} can be used to see if an entire
+ * line matches a Regex and output certain groups as a {@link KV}.
+ * </p>
+ * <p>
+ * {@link RegexTransform#find(String, int)} can be used to see if a portion of a line
+ * matches a Regex. {@link RegexTransform#matchesKV(String, int, int)} can be used to see if a
+ * portion of a line matches a Regex and output certain groups as a {@link KV}.
+ * </p>
+ * <p>
+ * Lines that do not match the Regex will not be output.
+ * </p>
+ */
+public class RegexTransform {
+  private RegexTransform() {
+    // do not instantiate
+  }
+
+  /**
+   * Returns a {@link RegexTransform.Matches} {@link PTransform} that checks if
+   * the entire line matches the Regex. Returns the entire line (group 0) as a
+   * {@link PCollection}.
+   * @param regex
+   *          The regular expression to run
+   */
+  public static Matches matches(String regex) {
+    return matches(regex, 0);
+  }
+
+  /**
+   * Returns a {@link RegexTransform.Matches} {@link PTransform} that checks if
+   * the entire line matches the Regex. Returns the group as a
+   * {@link PCollection}.
+   * @param regex
+   *          The regular expression to run
+   * @param group
+   *          The Regex group to return as a PCollection
+   */
+  public static Matches matches(String regex, int group) {
```

Comment:
What would say to a slightly more composition-oriented factoring?

 - `Matches` and `Find` as transforms that emit all groups as a `List<String>`.
 - The rest of the transforms here as that transform followed by `ParDo`.

It might look like this:

```java
class ProjectMatch {
  private final int group;

  public PCollection<String> apply(PCollection<String>) {
    return input.apply(Regex.matches(<regex>))
        // Whether or not to make the projection a composite transform wrapping a `ParDo`
        // or just `ParDo.of(Project.group(<n>)` is not terribly important.
        .apply(ParDo.of(Project.group(group));
  }
}

class MatchKv {
  private final int keyGroup, valueGroup;
  public PCollection<KV<String, String>> apply(PCollection<String>) {
    input.apply(Regex.matches(<regex>))
      .apply(Project.kvs(keyGroup, valueGroup))
  }
}
```

Here's my reasoning, so feel free to counter:

 - Regex logic isolated & composition utilized. It is somewhat a core tenet of Beam that you shouldn't have to hand-fuse in most cases.
 - You'd only need to customize the `DoFn` to avoid allocation of the list and extraneous groups. Since you are probably writing the regex yourself, you can simply avoid capturing groups where you don't want them, and this would be a very low-level optimization anyhow.
 - Supports somewhat more generic operation.
 - And these simple transforms are an easy reference for someone who actually needs that generic operation.